server/mmauth/mmauth.go

```diff
@@ -0,0 +1,249 @@
+	b := make([]byte, 16)
+	rand.Read(b)
+	securityToken := base64.URLEncoding.EncodeToString(b)
+	return securityToken
+}
+
+func deleteCookie(name string, w http.ResponseWriter) {
+	redirectCookie := http.Cookie{Name: name, Path: "/", Value: "", MaxAge: -1, Expires: time.Now().Add(-100 * time.Hour), HttpOnly: true}
+	http.SetCookie(w, &redirectCookie)
+}
+
+func (a *MMAuth) handleOAuthCallback(w http.ResponseWriter, r *http.Request) {
+	// Read oauthState from Cookie
+	stateCookie, err := r.Cookie("oauthstate")
+	if err != nil {
+		a.handleError("oauthstate cookie not found", err, w, r)
+		return
+	}
+
+	redirectCookie, err := r.Cookie("oauthredirect")
+	if err != nil {
+		a.handleError("oauthredirect cookie not found", err, w, r)
+		return
+	}
+
+	// Delete cookies
+	deleteCookie("oauthredirect", w)
+	deleteCookie("oauthstate", w)
+
+	if r.FormValue("state") != stateCookie.Value {
+		log.Println("invalid oauth state")
+		http.Redirect(w, r, "/", http.StatusTemporaryRedirect)
+		return
+	}
+
+	code := r.FormValue("code")
+	token, err := a.config.Exchange(context.Background(), code)
+	if err != nil {
+		a.handleError("code exchange error", err, w, r)
+		return
+	}
+
+	// Get user info from Mattermost
+	client := mattermost.NewAPIv4Client(a.params.MattermostURL)
+	client.SetOAuthToken(token.AccessToken)
+
+	etag := ""
+	user, resp := client.GetMe(etag)
+	if resp.Error != nil {
+		a.handleError("failed getting Mattermost user", resp.Error, w, r)
+		return
+	}
+
+	log.Printf("UserInfo: %v\n", user)
+
+	teams, resp := client.GetAllTeams(etag, 0, 10)
+	if resp.Error != nil {
+		a.handleError("failed getting Mattermost teams", err, w, r)
+		return
+	}
+
+	if len(teams) < 1 {
+		a.handleError("User has no teams", err, w, r)
+		return
+	}
+
+	log.Printf("%d team(s).", len(teams))
+	teamId := teams[0].Id
+
+	channels, resp := client.GetChannelsForTeamForUser(teamId, user.Id, false, etag)
+	if resp.Error != nil {
+		a.handleError("failed getting Mattermost channels for user", err, w, r)
+		return
+	}
+
+	if len(channels) < 1 {
+		a.handleError("User has no channels", err, w, r)
+		return
+	}
+
+	log.Printf("%d channel(s).", len(channels))
+	channelId := channels[0].Id
+
+	log.Printf("userId: %s, teamId: %s, channelId: %s", user.Id, teamId, channelId)
+
+	// Create native user
+	// TODO: Update user info with any changes?
+	nativeUser, _ := a.store.GetUserById(user.Id)
+	if nativeUser == nil {
+		err = a.store.CreateUser(&model.User{
+			ID:          user.Id,
+			Username:    user.Username,
+			Email:       user.Email,
+			Password:    "",
+			MfaSecret:   "",
+			AuthService: "mattermost",
+			AuthData:    "",
+			Props:       map[string]interface{}{},
+		})
+		if err != nil {
+			a.handleError("Unable to create the new user", err, w, r)
+			return
+		}
+	}
+
+	// Create session
+	// TODO: Remove any existing sessions with the same token? (Happens on re-login)
+	session := model.Session{
+		ID:     uuid.New().String(),
+		Token:  token.AccessToken,
```

Comment:
This is an interesting approach, using the the MM token as the authentication token for focalboard. My initial thought would be to create a focalboard specific session token and store the MM token in the session separately. Like the user ID above this will work but might get a little weird when using non-MM oauth service providers. I think this works more cleanly than the user ID above but you'd need to make sure that any requirements (eg length, characters) you have on the session token are loose enough to fit whatever pattern any oauth service provider might use for a token.

I can't think of any security reasons why this might be bad but @jupenur might have an opinion there

Reply:
This is probably fine for MM auth but for generic OAuth I'd definitely create a new token.